acc/components/optics/guide.py

```diff
@@ -62,57 +58,43 @@ def propagate(
         ch1 = -hh1*l - z*hh; ch2 = y*l
         # Compute intersection times.
         t1 = (l - z) / vz  # for guide exit
-        i = 0
-        if vdotn_v1 < 0:
-            t2 = (cv1 - cv2)/vdotn_v1
-            if t2 < t1:
-                t1 = t2
-                i = 1
-        if vdotn_v2 < 0:
-            t2 = (cv1 + cv2)/vdotn_v2
-            if t2<t1:
-                t1 = t2
-                i = 2
-        if vdotn_h1 < 0:
-            t2 = (ch1 - ch2)/vdotn_h1
-            if t2<t1:
-                t1 = t2
-                i = 3
-        if vdotn_h2 < 0:
-            t2 = (ch1 + ch2)/vdotn_h2
-            if t2 < t1:
-                t1 = t2
-                i = 4
-        if i == 0:
-            break                    # Neutron left guide.
+        vdots = (-1., vdotn_v1, vdotn_v2, vdotn_h1, vdotn_h2)
+        times = (t1,
+                 (cv1 - cv2) / vdotn_v1,
+                 (cv1 + cv2) / vdotn_v2,
+                 (ch1 - ch2) / vdotn_h1,
+                 (ch1 + ch2) / vdotn_h2)
+        best_time = inf
+        for i in range(0, 5):
+            if vdots[i] < 0 and best_time > times[i]:
+                best_time = times[i]
+                side = i
```

Comment:
So, here, if I try,
```python
is_better = vdots[i] < 0 and best_time > times[i]
if is_better:
    best_time = times[i]
    side = i
```
then it's still (relatively) fast but with,
```python
is_better = vdots[i] < 0 and best_time > times[i]
best_time = is_better and times[i] or best_time
side = is_better and i or side
```
it becomes extremely slow. Have you any tips, @ckendrick?

Reply:
There might be an issue with how `side` is being set. I ran into an issue where I was setting the side to an incorrect value which caused neutrons to get stuck in the loop for all `max_bounces` number of iterations. If you set `max_bounces = 100` and it is significantly faster, then this is probably the issue. Also, running `test_guide.py` in interactive mode should show differences if the side is incorrect.

Sorry I haven't opened a PR yet for the optimized tapered guide. The equivalent loop from mine is
```
i = 0
for k in range(4):
    side_hit = vdots[k] < 0.0
    t2 = times[k] / vdots[k]
    best_time = side_hit and t2 < t1
    i += best_time * int(k + 1)
    t1 = ((not best_time) * t1) + best_time * t2
if i == 0:
    break
```

Reply:
Aha, thank you. Nice trick with using the numerical aspect of the truth value more directly than I.

And, yes, reducing `max_bounces` correspondingly speeds the run. Maybe my two alternatives aren't logically equivalent and I've yet to spot why!